### PR TITLE
[draft] AMP Phase 6: MINLPTests, performance, and docs

### PR DESCRIPTION
**Pull request by bernalde**
## What changed
Adds the scoped checklist for the validation, performance, and documentation slice.

## Why
This work should happen once the solver is feature-complete enough to benchmark and document against MINLPTests and Alpine.

## Implementation plan
- merge or rebase the MINLPTests infrastructure into the AMP line of work
- register AMP and track known failures explicitly
- add warm starts, sparse matrices, logging, and notebook/documentation follow-through

## Validation
- Reviewed the staged tracking doc locally
- Branch is pushed and ready for follow-up commits

### CODE REVIEW

**Comment by bernalde** [comment]
MINLPTests baseline is now available from the synced Phase 1 branch.

Using the upstream MINLPTests infrastructure merged in `upstream/main` (`f7af559`), the current AMP implementation baseline is:
- `nlp`: 7 / 15 passed
- `nlp_mi`: 1 / 13 passed
- `infeasible`: 2 / 3 passed

Failure shape right now:
- many false-infeasible results on nonconvex NLP and MINLP cases
- several wrong objectives on MINLP cases
- one false-feasible infeasible case (`nlp_mi_007_020`)

That benchmark summary is documented in more detail on PR #9. This PR can now treat that result as the starting point for Phase 9/10 work rather than needing to invent separate infrastructure.

**Review by bernalde — COMMENTED** [review]
I reviewed the PR head locally in `/tmp/discopt-pr14`, compared it against both `feature/amp-completion/roadmap` and the current AMP Phase 5 line, and ran `~/.local/bin/uv run pytest python/tests/test_amp.py -m "not slow"` from the PR worktree. That run reported `2 failed, 61 passed, 1 deselected, 1 xfailed`.

The main blockers are:
- this branch is still stacked on the earlier AMP scope commits instead of only showing Phase 6 work
- the diff does not contain the MINLPTests, benchmark, known-failure, or notebook changes that the Phase 6 scope file describes
- two of the new tests are already stale against the current AMP line

I recorded the specific changes I recommend inline below.

File: plans/amp-completion/01-robustness.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# AMP Phase 1: Bug Fixes and Robustness
```
Comment (bernalde) [inline plans/amp-completion/01-robustness.md:1]: This branch is still carrying the earlier AMP scope commits (`9be9376` through `dbfe515`) on top of `feature/amp-completion/roadmap`. Since Phases 1-5 have already been reviewed separately, this PR needs to be rebased or restacked onto the current AMP base so the diff is limited to Phase 6 work. Right now the review surface is dominated by old plan files and stale solver/test changes that are unrelated to the stated PR scope.

File: plans/amp-completion/06-benchmarks-polish.md [hunk 2]
```diff
@@ -0,0 +1,18 @@
+# AMP Phase 6: Validation, Performance, and Documentation
+
+Plan phases:
+- 9 MINLPTests integration
+- 10 Performance and polish
+
+Primary files:
```
Comment (bernalde) [inline plans/amp-completion/06-benchmarks-polish.md]: This scope file lists `python/tests/test_minlptests.py`, `python/tests/data/known_failures.toml`, `discopt_benchmarks/benchmarks/problems/minlptests_problems.py`, and the notebook as the primary files for this PR, but none of those files are actually in the diff. As written, the PR does not implement Phase 6 yet; it only adds this checklist plus older AMP commits. Either narrow the PR to just tracking the Phase 6 plan file or include the MINLPTests, benchmark, and documentation changes that this scope describes.

File: python/tests/test_amp.py [hunk 3]
```diff
@@ -760,6 +760,37 @@ def test_piecewise_interval_rows_use_interval_specific_bounds(self):
...
+class TestAmpPhase1Helpers:
+    """Fast regression tests for Phase 1 helper behavior."""
+
+    def test_piecewise_big_m_scales_with_large_coefficients(self):
+        """Big-M must scale with coefficient magnitude instead of adding a flat 1.0."""
+        from discopt._jax.milp_relaxation import _compute_piecewise_big_m
+
+        corners = [-25000.0, 10000.0, 5000.0, 30000.0]
+        big_m = _compute_piecewise_big_m(corners)
+        expected = 30000.0 * (1.0 + 1e-4) + 1e-2
```
Comment (bernalde) [inline python/tests/test_amp.py]: This expectation is stale against the current AMP line. `_compute_piecewise_big_m` now uses a magnitude-proportional floor, so after rebasing onto the already-merged AMP work this value is `30006.0`, not `30003.01`. That is one of the failures I hit in the local `python/tests/test_amp.py -m "not slow"` run. Please update the expected value when this branch is restacked.

File: python/tests/test_amp.py [hunk 4]
```diff
@@ -1014,6 +1062,64 @@ def test_solve_model_signature_exposes_solver_parameter(self):
...
+            call_sizes.append(size)
+            status = "infeasible" if size >= 4 else "optimal"
+            return FakeMilpModel(status), {"dummy": True}
+
+        monkeypatch.setattr(
+            "discopt._jax.milp_relaxation.build_milp_relaxation",
+            fake_build,
+        )
+
+        result, _, kept_cuts = amp_mod._solve_milp_with_oa_recovery(
```
Comment (bernalde) [inline python/tests/test_amp.py:1889]: This test is also behind the current AMP API. On the current solver path `_solve_milp_with_oa_recovery` threads `convhull_formulation`, `convhull_ebd`, and `convhull_ebd_encoding`, and the fake builder has to accept the same arguments. After rebasing, this call fails before it can exercise the cut-dropping logic, so the test needs to be updated along with the helper signature.

File: python/discopt/solvers/amp.py [hunk 5]
```diff
@@ -108,16 +109,146 @@ def _check_integer_feasible(
...
+    incumbent: Optional[np.ndarray],
+    oa_cuts: Optional[list],
+    time_limit: Optional[float],
+    gap_tolerance: float,
+):
+    """Retry MILP solves after dropping the oldest half of OA cuts on infeasibility."""
+    from discopt._jax.milp_relaxation import build_milp_relaxation
+
+    active_oa_cuts = list(oa_cuts or [])
+    while True:
```
Comment (bernalde) [inline python/discopt/solvers/amp.py]: This `while True:` loop terminates only because `active_oa_cuts` is halved each iteration until it empties, at which point the early-return fires. But the termination guarantee is implicit and fragile: any future change that makes `drop_count` zero (say, rounding down from 1 cut) would spin forever.

Replace with a bounded loop:
```python
max_retries = max(1, len(active_oa_cuts).bit_length() + 1)
for _retry in range(max_retries):
```
The bit-length bound is O(log N) and makes the worst-case explicit.

**Review by bernalde — COMMENTED** [review]
Detailed review of the Phase 1 robustness, maximize handling, and planning changes. All 93 non-slow tests pass. The architecture of the changes is sound: the minimize/maximize unification via `_from_minimization_space` is clean, Big-M scaling is an important correctness fix, and the integer rounding fallback is a meaningful improvement over simple `round()`. Several items warrant attention before merging, ranging from a potential infinite loop to a missed integrality guard and some dead code.

File: python/discopt/solvers/amp.py [hunk 6]
```diff
@@ -342,43 +485,44 @@ def solve_amp(
...
                 constraint_ub,
-            )
-            if feasible and obj_nlp < UB:
-                UB = obj_nlp
+            ):
+                continue
+            x_nlp = cand_x
+            obj_nlp = cand_obj
+            obj_nlp_min = float(cand_obj)
+            break
```
Comment (bernalde) [inline python/discopt/solvers/amp.py]: This `break` accepts the first feasible NLP candidate and skips all remaining rounding alternatives. When the NLP starting point from `floor` or `ceil` leads to a substantially better objective than the nearest-round starting point, the solver misses that incumbent.

Consider tracking the best feasible candidate across the full loop, similar to:
```python
if best_obj is None or float(cand_obj) < best_obj:
    best_x, best_obj = cand_x, float(cand_obj)
```
If runtime is a concern, an early-exit threshold (e.g., break when the objective is within `abs_tol` of the current LB) would be a reasonable compromise.

File: python/discopt/solvers/amp.py [hunk 7]
```diff
@@ -342,43 +485,44 @@ def solve_amp(
...
                 evaluator,
-                x_nlp,
+                cand_x,
                 constraint_lb,
                 constraint_ub,
-            )
-            if feasible and obj_nlp < UB:
-                UB = obj_nlp
+            ):
+                continue
```
Comment (bernalde) [inline python/discopt/solvers/amp.py]: The NLP candidate loop checks constraint feasibility with `_check_constraints_with_evaluator` but does not verify that integer variables in the NLP solution are actually integral. The NLP solver receives fixed bounds for integer variables via `_build_fixed_integer_bounds`, but solver tolerances can still produce slightly fractional values in the returned `cand_x`.

Adding `if not _check_integer_feasible(cand_x, model): continue` before accepting the candidate would prevent false incumbents from leaking into the upper bound. The function already exists and is used elsewhere in this file.

File: python/discopt/solvers/amp.py [hunk 8]
```diff
@@ -242,11 +373,19 @@ def solve_amp(
...
     from discopt._jax.nlp_evaluator import NLPEvaluator
-    from discopt._jax.milp_relaxation import build_milp_relaxation
     from discopt._jax.partition_selection import pick_partition_vars
     from discopt._jax.term_classifier import classify_nonlinear_terms
     from discopt.solvers.nlp_ipopt import _infer_constraint_bounds
 
+    assert model._objective is not None
+    maximize = model._objective.sense == ObjectiveSense.MAXIMIZE
+
+    def _to_minimization_space(value: float) -> float:
```
Comment (bernalde) [inline python/discopt/solvers/amp.py]: `_to_minimization_space` is defined but never called. Both `build_milp_relaxation` and `NLPEvaluator` already negate the objective for maximize problems internally, so no explicit conversion is needed by the caller. Either remove this dead function, or add a short comment explaining why it exists but is intentionally unused (e.g., as documentation of the design contract).

File: python/discopt/solvers/amp.py [hunk 9]
```diff
@@ -342,43 +485,44 @@ def solve_amp(
...
+                cand_x,
                 constraint_lb,
                 constraint_ub,
-            )
-            if feasible and obj_nlp < UB:
-                UB = obj_nlp
+            ):
+                continue
+            x_nlp = cand_x
+            obj_nlp = cand_obj
```
Comment (bernalde) [inline python/discopt/solvers/amp.py]: `obj_nlp` is assigned here but never read after the loop exits. Only `obj_nlp_min` is used in the incumbent comparison (line 518). The three-way None check on line 516 (`if x_nlp is not None and obj_nlp is not None and obj_nlp_min is not None`) can be simplified to just `x_nlp` and `obj_nlp_min`, removing the redundant variable.

File: python/discopt/_jax/milp_relaxation.py [hunk 10]
```diff
@@ -139,6 +140,12 @@ def _piecewise_product_bounds(
     return corners, min(corners), max(corners)
 
 
+def _compute_piecewise_big_m(corners: list[float]) -> float:
+    """Scale Big-M with the interval magnitude instead of adding a flat constant."""
+    max_corner = max(abs(float(c)) for c in corners)
+    return max_corner * (1.0 + 1e-4) + 1e-2
```
Comment (bernalde) [inline python/discopt/_jax/milp_relaxation.py]: The fixed `+ 1e-2` floor is fine for moderately-sized intervals, but it can be disproportionately large for near-zero product ranges. For instance, if all corners are within `[-0.001, 0.001]`, the Big-M becomes roughly 0.01 -- an order of magnitude larger than the feasible range.

A proportional floor like `max(1e-6, 1e-4 * max_corner)` keeps the Big-M tight across scales:
```python
return max_corner * (1.0 + 1e-4) + max(1e-6, 1e-4 * max_corner)
```
This matches the scaling philosophy already present in the `(1.0 + 1e-4)` multiplier.

File: python/discopt/solvers/amp.py [hunk 11]
```diff
@@ -108,16 +109,146 @@ def _check_integer_feasible(
...
+
+    offset = 0
+    for v in model._variables:
+        if v.var_type in (VarType.BINARY, VarType.INTEGER):
+            v_lb = np.asarray(v.lb, dtype=np.float64).ravel()
+            v_ub = np.asarray(v.ub, dtype=np.float64).ravel()
+            for k in range(v.size):
+                idx = offset + k
+                val = float(np.clip(x[idx], v_lb[k], v_ub[k]))
+                rounded = round(val)
```
Comment (bernalde) [inline python/discopt/solvers/amp.py]: `round(val)` can produce a value outside `[v_lb[k], v_ub[k]]` when the variable bound is not an integer. For example, if `v_ub[k] = 2.6`, then `val = 2.6` after clipping, and `round(2.6) = 3`, which exceeds the upper bound.

For well-formed MINLP models integer bounds are integral, so this is an edge case. But a defensive clamp after rounding is cheap:
```python
rounded = int(round(val))
rounded = min(max(rounded, int(np.ceil(v_lb[k]))), int(np.floor(v_ub[k])))
```

File: python/discopt/solvers/amp.py [hunk 12]
```diff
@@ -108,16 +109,146 @@ def _check_integer_feasible(
...
+
+    candidates: list[np.ndarray] = []
+    if total_candidates <= max_candidates:
+        option_lists = [options for _, options in integer_entries]
+        for values in itertools.product(*option_lists):
+            cand = base.copy()
+            for (idx, _), value in zip(integer_entries, values):
+                cand[idx] = float(value)
+            candidates.append(cand)
+    else:
```
Comment (bernalde) [inline python/discopt/solvers/amp.py:279]: When `total_candidates > max_candidates`, the fallback generates `1 + sum(len(options) - 1)` candidates without checking against `max_candidates`. With many integer variables (each having 2-3 options), this can still produce a sizeable set.

For example, 30 binary variables with 2 options each: fallback = 1 + 30 = 31, which is fine. But the fallback is linear in the number of integer variables rather than bounded, and in a model with 100+ integer variables each having 3 options, it produces 201 candidates, none of which are explored by the combination product.

Consider capping the output: `return deduped[:max_candidates]`.

File: python/tests/test_amp.py [hunk 13]
```diff
@@ -1014,6 +1062,64 @@ def test_solve_model_signature_exposes_solver_parameter(self):
...
+                self.status = status
+
+            def solve(self, time_limit=None, gap_tolerance=None):
+                return MilpRelaxationResult(
+                    status=self.status,
+                    objective=0.0,
+                    x=np.zeros(1, dtype=np.float64),
+                )
+
+        def fake_build(model, terms, disc_state, incumbent, oa_cuts=None):
```
Comment (bernalde) [inline python/tests/test_amp.py]: The mock signature `def fake_build(model, terms, disc_state, incumbent, oa_cuts=None)` matches the current `_solve_milp_with_oa_recovery` caller, but the real `build_milp_relaxation` already accepts additional keyword arguments (`convhull_formulation`, `convhull_ebd`, `convhull_ebd_encoding` from Phase 5). When those are wired through to the recovery function, this test will break.

Adding `**kwargs` to the mock signature would make it forward-compatible:
```python
def fake_build(model, terms, disc_state, incumbent, oa_cuts=None, **kwargs):
```

File: plans/amp-completion/05-advanced-formulations.md [hunk 14]
```diff
@@ -0,0 +1,18 @@
+# AMP Phase 5: Advanced Formulations and Adaptive Partitioning
+
+Plan phases:
+- 7 Facet formulation and embedded binary encoding
+- 8 Adaptive variable selection
+
+Primary files:
+- `python/discopt/_jax/milp_relaxation.py`
+- `python/discopt/_jax/embedding.py`
+- `python/discopt/_jax/partition_selection.py`
```
Comment (bernalde) [inline plans/amp-completion/05-advanced-formulations.md:10]: Files `python/discopt/_jax/embedding.py` and `python/discopt/_jax/partition_selection.py` are listed here. `partition_selection.py` already exists (it was added in an earlier phase), while `embedding.py` does not exist yet. Marking which files are existing vs. to-be-created would help readers gauge implementation progress. The same applies to "Primary files" sections in the other plan documents.

File: plans/amp-completion/06-benchmarks-polish.md (hunk 2 above)
Comment (bernalde) [reply plans/amp-completion/06-benchmarks-polish.md]: Addressed in [45967ed](https://github.com/bernalde/discopt/commit/45967ed0e8a3c0c56299065f256f2139272cc5f8). I narrowed the branch to a planning-only Phase 6 PR and updated [06-benchmarks-polish.md](https://github.com/bernalde/discopt/blob/45967ed0e8a3c0c56299065f256f2139272cc5f8/plans/amp-completion/06-benchmarks-polish.md#L7-L15) to mark the touchpoints as existing vs. to-be-added and to say explicitly that this PR only adds the scoped checklist.

File: plans/amp-completion/01-robustness.md (hunk 1 above)
Comment (bernalde) [reply plans/amp-completion/01-robustness.md:1]: Addressed by restacking the branch in [60eb64d](https://github.com/bernalde/discopt/commit/60eb64dd8c618a98025430577d75149d77560f9a) and retargeting the PR base to `feature/amp-completion/phase-5-advanced-formulations`. The current review surface is now limited to [06-benchmarks-polish.md](https://github.com/bernalde/discopt/blob/45967ed0e8a3c0c56299065f256f2139272cc5f8/plans/amp-completion/06-benchmarks-polish.md#L1-L21), [amp.py](https://github.com/bernalde/discopt/blob/45967ed0e8a3c0c56299065f256f2139272cc5f8/python/discopt/solvers/amp.py), and [test_amp.py](https://github.com/bernalde/discopt/blob/45967ed0e8a3c0c56299065f256f2139272cc5f8/python/tests/test_amp.py).

File: python/tests/test_amp.py (hunk 3 above)
Comment (bernalde) [reply python/tests/test_amp.py]: Handled by the restack in [60eb64d](https://github.com/bernalde/discopt/commit/60eb64dd8c618a98025430577d75149d77560f9a). The stale expectation dropped out with the old Phase 1 diff, and the current regression in [test_piecewise_big_m_scales_with_large_coefficients](https://github.com/bernalde/discopt/blob/45967ed0e8a3c0c56299065f256f2139272cc5f8/python/tests/test_amp.py#L1099-L1108) now matches the proportional-floor behavior from the AMP base.

File: python/tests/test_amp.py (hunk 4 above)
Comment (bernalde) [reply python/tests/test_amp.py:1889]: Handled by the restack in [60eb64d](https://github.com/bernalde/discopt/commit/60eb64dd8c618a98025430577d75149d77560f9a). The current OA recovery test now passes the Phase 5 convex-hull keywords through the fake builder at [test_oa_cut_recovery_drops_oldest_half](https://github.com/bernalde/discopt/blob/45967ed0e8a3c0c56299065f256f2139272cc5f8/python/tests/test_amp.py#L1680-L1714), so it exercises the cut-recovery path again.

File: python/discopt/solvers/amp.py (hunk 5 above)
Comment (bernalde) [reply python/discopt/solvers/amp.py]: Handled by the restack in [60eb64d](https://github.com/bernalde/discopt/commit/60eb64dd8c618a98025430577d75149d77560f9a). `_solve_milp_with_oa_recovery` now uses an explicit bounded retry loop at [amp.py](https://github.com/bernalde/discopt/blob/45967ed0e8a3c0c56299065f256f2139272cc5f8/python/discopt/solvers/amp.py#L276-L309), and [test_oa_cut_recovery_drops_oldest_half](https://github.com/bernalde/discopt/blob/45967ed0e8a3c0c56299065f256f2139272cc5f8/python/tests/test_amp.py#L1662-L1718) still covers the retry path.

File: python/discopt/solvers/amp.py (hunk 6 above)
Comment (bernalde) [reply python/discopt/solvers/amp.py]: Handled by the restack in [60eb64d](https://github.com/bernalde/discopt/commit/60eb64dd8c618a98025430577d75149d77560f9a). `_solve_best_nlp_candidate` now scans the full rounding set and keeps the lowest feasible objective at [amp.py](https://github.com/bernalde/discopt/blob/45967ed0e8a3c0c56299065f256f2139272cc5f8/python/discopt/solvers/amp.py#L232-L258). The regression is covered by [test_best_nlp_candidate_chooses_lowest_feasible_objective](https://github.com/bernalde/discopt/blob/45967ed0e8a3c0c56299065f256f2139272cc5f8/python/tests/test_amp.py#L1569-L1620).

File: python/discopt/solvers/amp.py (hunk 7 above)
Comment (bernalde) [reply python/discopt/solvers/amp.py]: Handled by the restack in [60eb64d](https://github.com/bernalde/discopt/commit/60eb64dd8c618a98025430577d75149d77560f9a). `_check_integer_feasible` is now applied before accepting an NLP candidate at [amp.py](https://github.com/bernalde/discopt/blob/45967ed0e8a3c0c56299065f256f2139272cc5f8/python/discopt/solvers/amp.py#L241-L245), and [test_best_nlp_candidate_rejects_noninteger_nlp_return](https://github.com/bernalde/discopt/blob/45967ed0e8a3c0c56299065f256f2139272cc5f8/python/tests/test_amp.py#L1622-L1660) covers the case.

File: python/discopt/solvers/amp.py (hunk 8 above)
Comment (bernalde) [reply python/discopt/solvers/amp.py]: Addressed in [45967ed](https://github.com/bernalde/discopt/commit/45967ed0e8a3c0c56299065f256f2139272cc5f8). I removed the unused `_to_minimization_space` helper; `solve_amp` now only keeps `_from_minimization_space` for the caller-facing sign conversion at [amp.py](https://github.com/bernalde/discopt/blob/45967ed0e8a3c0c56299065f256f2139272cc5f8/python/discopt/solvers/amp.py#L533-L539).

File: python/discopt/solvers/amp.py (hunk 9 above)
Comment (bernalde) [reply python/discopt/solvers/amp.py]: Handled by the restack in [60eb64d](https://github.com/bernalde/discopt/commit/60eb64dd8c618a98025430577d75149d77560f9a). The redundant `obj_nlp` variable is gone in the current `_solve_best_nlp_candidate` flow, which now only tracks `best_x` and `best_obj` at [amp.py](https://github.com/bernalde/discopt/blob/45967ed0e8a3c0c56299065f256f2139272cc5f8/python/discopt/solvers/amp.py#L229-L258).

File: python/discopt/_jax/milp_relaxation.py (hunk 10 above)
Comment (bernalde) [reply python/discopt/_jax/milp_relaxation.py]: Handled by the restack in [60eb64d](https://github.com/bernalde/discopt/commit/60eb64dd8c618a98025430577d75149d77560f9a). The proportional Big-M floor already lives in the Phase 5 base now, and the current regression checks on this branch track that behavior at [test_amp.py](https://github.com/bernalde/discopt/blob/45967ed0e8a3c0c56299065f256f2139272cc5f8/python/tests/test_amp.py#L1099-L1117). No additional change was needed in this PR once the old diff dropped out.

File: python/discopt/solvers/amp.py (hunk 11 above)
Comment (bernalde) [reply python/discopt/solvers/amp.py]: Addressed in [45967ed](https://github.com/bernalde/discopt/commit/45967ed0e8a3c0c56299065f256f2139272cc5f8). `_build_fixed_integer_bounds` now clamps the rounded value to the realizable integer interval at [amp.py](https://github.com/bernalde/discopt/blob/45967ed0e8a3c0c56299065f256f2139272cc5f8/python/discopt/solvers/amp.py#L205-L212), and [test_build_fixed_integer_bounds_clamps_to_integer_domain](https://github.com/bernalde/discopt/blob/45967ed0e8a3c0c56299065f256f2139272cc5f8/python/tests/test_amp.py#L1522-L1537) covers the edge case.

File: python/discopt/solvers/amp.py (hunk 12 above)
Comment (bernalde) [reply python/discopt/solvers/amp.py:279]: Addressed in [45967ed](https://github.com/bernalde/discopt/commit/45967ed0e8a3c0c56299065f256f2139272cc5f8). `_integer_rounding_candidates` now returns `deduped[:max_candidates]` at [amp.py](https://github.com/bernalde/discopt/blob/45967ed0e8a3c0c56299065f256f2139272cc5f8/python/discopt/solvers/amp.py#L173-L180), and [test_integer_rounding_candidates_respect_max_candidates](https://github.com/bernalde/discopt/blob/45967ed0e8a3c0c56299065f256f2139272cc5f8/python/tests/test_amp.py#L1509-L1520) covers the cap.

File: python/tests/test_amp.py (hunk 13 above)
Comment (bernalde) [reply python/tests/test_amp.py]: Handled by the restack in [60eb64d](https://github.com/bernalde/discopt/commit/60eb64dd8c618a98025430577d75149d77560f9a). The current fake builder threads the convex-hull keyword arguments explicitly at [test_amp.py](https://github.com/bernalde/discopt/blob/45967ed0e8a3c0c56299065f256f2139272cc5f8/python/tests/test_amp.py#L1680-L1689), so the OA recovery test stays aligned with the Phase 5 caller.

File: plans/amp-completion/05-advanced-formulations.md (hunk 14 above)
Comment (bernalde) [reply plans/amp-completion/05-advanced-formulations.md:10]: Partially addressed in [45967ed](https://github.com/bernalde/discopt/commit/45967ed0e8a3c0c56299065f256f2139272cc5f8). I applied the existing/to-be-added labeling to the only remaining planning file in this PR, [06-benchmarks-polish.md](https://github.com/bernalde/discopt/blob/45967ed0e8a3c0c56299065f256f2139272cc5f8/plans/amp-completion/06-benchmarks-polish.md#L7-L15). I did not reopen the earlier phase planning docs from this branch because the PR has been restacked to Phase 6 only, so those files are now outside the review surface.

**Comment by bernalde** [comment]
Follow-up to the review summaries above: I restacked this PR onto `feature/amp-completion/phase-5-advanced-formulations`, narrowed it to the Phase 6 planning surface, and applied the remaining AMP rounding guard cleanups in [45967ed](https://github.com/bernalde/discopt/commit/45967ed0e8a3c0c56299065f256f2139272cc5f8). Validation on the rebased worktree used `PYTHONPATH=python ~/.local/bin/uv run pytest python/tests/test_amp.py -m "not slow"` and passed with `95 passed, 1 deselected, 1 xfailed`.

**Comment by bernalde** [comment]
I pushed [05aef50](https://github.com/bernalde/discopt/commit/05aef50) and [4c2e45c](https://github.com/bernalde/discopt/commit/4c2e45c) to this branch.

The latest completed full translated-suite comparison artifact still reports `alpine_only_pass=2`, `discopt_only_pass=11`, `both_fail=109`, and `both_pass=0`, with the Alpine-only cases listed as `nlp_cvx_001_010` and `nlp_cvx_002_010`. That artifact is now stale for those two LPs: both regressions now use the shared MINLPTests translations directly and pass on this branch through both the default solver path and `solver="amp"`.

The AMP regression tests that depend on translated MINLPTests mixed-integer cases also now pull those instances from `python/tests/test_minlptests.py` instead of duplicating the models locally. I also verified that the repo `.venv` has `cyipopt` importable, so the fixed-integer Ipopt retry path is available in the local environment we use for the MINLPTests work.

For follow-up scope, I opened #15, #16, #17, and #18 for the Alpine functionality that AMP still does not implement, and #19 as the tracker for any remaining Alpine-only MINLPTests wins once the full comparison is refreshed.

**Comment by bernalde** [comment]
I pushed [75051f2](https://github.com/bernalde/discopt/commit/75051f2c3a6c969ea5c50a35ef61e456e6931d74), which updates the Phase 6 benchmark notes in [06-benchmarks-polish.md](https://github.com/bernalde/discopt/blob/75051f2c3a6c969ea5c50a35ef61e456e6931d74/plans/amp-completion/06-benchmarks-polish.md), reconciles the `nlp_mi` status note in [07-minlptests-remediation.md](https://github.com/bernalde/discopt/blob/75051f2c3a6c969ea5c50a35ef61e456e6931d74/plans/amp-completion/07-minlptests-remediation.md), and keeps the Alpine-only rerun mode in [run_full_minlptests_benchmark.sh](https://github.com/bernalde/discopt/blob/75051f2c3a6c969ea5c50a35ef61e456e6931d74/scripts/run_full_minlptests_benchmark.sh).

Using the completed full translated-suite comparison from the local run, the current head-to-head status is:
- discopt AMP: `85/122` pass (`62/91` `nlp_cvx`, `9/15` `nlp`, `11/13` `nlp_mi`, `3/3` infeasible)
- Alpine.jl: `2/122` pass (`nlp_cvx_001_010`, `nlp_cvx_002_010`)
- split: `both_pass=2`, `discopt_only_pass=83`, `alpine_only_pass=0`, `both_fail=37`

So the refreshed full run leaves no Alpine-only translated MINLPTests wins. The remaining benchmark work is now entirely on the discopt side:
- convex false infeasibility, especially `nlp_cvx_501_011_{1d..20d}` and the `108_*` / `203_*` families
- one convex objective miss on `nlp_cvx_106_010`
- six nonconvex `nlp` misses
- two `nlp_mi` cases, `nlp_mi_003_014` and `nlp_mi_003_015`, that recover objective `11.0` but still exit as `time_limit`

Issue [#19](https://github.com/bernalde/discopt/issues/19) has been updated from this artifact and can now be treated as completed. The remaining follow-up scope stays in [#15](https://github.com/bernalde/discopt/issues/15) through [#18](https://github.com/bernalde/discopt/issues/18).

**Comment by bernalde** [comment]
I pushed [ead7a0b](https://github.com/bernalde/discopt/commit/ead7a0bd7d0a44e1d930bd0ac673652be0fbc65d) to remove the machine-specific benchmark defaults from [run_full_minlptests_benchmark.sh](https://github.com/bernalde/discopt/blob/ead7a0bd7d0a44e1d930bd0ac673652be0fbc65d/scripts/run_full_minlptests_benchmark.sh).

The script now only uses:
- repo-relative defaults, `../Alpine.jl` and `../MINLPTests.jl`
- explicit `ALPINE_PROJECT` and `MINLPTESTS_PATH` overrides

There are no remaining `/home/bernalde/...` or similar workstation-specific paths in the Phase 6 runner or the updated benchmark notes.
